tools/chainsync-loadtest/src/concurrency/scope.rs

```
@@ -0,0 +1,184 @@
+            if is_empty {
+                return;
+            } else {
+                n.await;
+            }
+        }
+    }
+    pub fn inc(&self) {
+        *self.count.lock().unwrap() += 1;
+    }
+    pub fn dec(&self) {
+        let mut count = self.count.lock().unwrap();
+        *count -= 1;
+        if *count != 0 {
+            return;
+        }
+        self.empty.notify_waiters();
+    }
+}
+
+struct ScopeState {
+    err: Option<anyhow::Error>,
+    cancelled: bool,
+}
+
+// Scope represents a collection of spawned futures with a common bounded lifetime.
+// See Scope::run for details.
+pub struct Scope {
+    ctx: CtxWithCancel,
+    // Main futures have to exit before the scope is cancelled.
+    main_futures: WaitGroup,
+    // Additional futures running in the background,
+    // which will be cancelled as soon as main_futures exit.
+    weak_futures: WaitGroup,
+    state: Mutex<ScopeState>,
+}
+
+impl Scope {
+    fn complete(&self, v: anyhow::Result<()>) {
+        let e = if let Err(e) = v {
+            e
+        } else {
+            return;
+        };
+        let mut s = self.state.lock().unwrap();
+        if s.err.is_some() {
+            return;
+        }
+        // Ignore Cancelled errors after the whole scope has been cancelled.
+        if s.cancelled
+            && e.downcast_ref::<ctx::Error>().map(|e| e == &ctx::Error::Cancelled).unwrap_or(false)
+        {
+            return;
+        }
+        s.err = Some(e);
+        s.cancelled = true;
+        self.ctx.cancel();
+    }
+
+    // Spawn a "main" future in the scope.
+    pub fn spawn<F>(self: &Arc<Self>, f: impl Send + FnOnce(Ctx, Arc<Self>) -> F)
+    where
+        F: Future<Output = anyhow::Result<()>> + Send + 'static,
+    {
+        let fut = f((*self.ctx).clone(), self.clone());
+        let s = self.clone();
+        s.main_futures.inc();
+        tokio::spawn(async move {
+            s.complete(fut.await);
+            s.main_futures.dec();
```

This is suspect -- if `fut` panics, we don't do `dec` (and, in general, we swallow the panic).